cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs

```diff
@@ -1422,6 +1424,33 @@ toTentativeEpochInfo (EraHistory _ interpreter) =
     $ hoistEpochInfo (first (Text.pack . show) . runExcept)
     $ Consensus.interpreterToEpochInfo (Consensus.unsafeExtendSafeZone interpreter)
 
+
+-- | Get slot number for timestamp, or an error if the UTC timestamp is before 'SystemStart'
+utcTimeToSlotNo
+  :: SocketPath
+  -> AnyConsensusModeParams
+  -> NetworkId
+  -> UTCTime
+  -> ExceptT ShelleyQueryCmdError IO SlotNo
+utcTimeToSlotNo (SocketPath sockPath) (AnyConsensusModeParams cModeParams) network utcTime = do
+  let localNodeConnInfo = LocalNodeConnectInfo cModeParams network sockPath
+  case consensusModeOnly cModeParams of
+    CardanoMode -> do
+      (systemStart, eraHistory) <- executeLocalStateQueryExpr' localNodeConnInfo $
+        (,) <$> queryExpr' QuerySystemStart
+            <*> queryExpr' (QueryEraHistory CardanoModeIsMultiEra)
+      let relTime = toRelativeTime systemStart utcTime
+      hoistEither $ Api.getSlotForRelativeTime relTime eraHistory & first ShelleyQueryCmdPastHorizon
```

Comment:
You can convert the above applicative into a `do` expression then move the definition of `relTime` and call to `getSlotForRelativeTime` into the `do` expression.  This will save you from having to marshal the data out via a tuple.

Reply:
Something like this:

```
    executeLocalStateQueryExpr' localNodeConnInfo $ do
      systemStart <- queryExpr' QuerySystemStart
      eraHistory <- queryExpr' (QueryEraHistory CardanoModeIsMultiEra)
      let relTime = toRelativeTime systemStart utcTime
      hoistEither $ Api.getSlotForRelativeTime relTime eraHistory & first ShelleyQueryCmdPastHorizon
```

Reply:
Might have to tweak the error handling for this though.

Reply:
I'm not sure about that. On the one side we would avoid using tuple here, but on the other side it slightly obfuscates what's happening inside. I wanted to limit the scope of `ExceptT ShelleyQueryCmdError (LocalStateQueryExpr (BlockInMode mode) ChainPoint (QueryInMode mode) () IO) a` as much as possible to make reasoning easier.